apollo-federation/src/sources/connect/json_selection/apply_to.rs

```diff
@@ -148,27 +155,60 @@ impl ApplyToError {
     }
 }
 
+// Rust doesn't allow implementing methods directly on tuples like
+// (Option<JSON>, Vec<ApplyToError>), so we define a trait to provide the
+// methods we need, and implement the trait for the tuple in question.
+pub(super) trait ApplyToResultMethods {
+    fn prepend_errors(self, errors: &[ApplyToError]) -> Self;
+
+    fn and_then_collecting_errors(
+        &self,
+        f: impl FnOnce(&JSON) -> (Option<JSON>, Vec<ApplyToError>),
+    ) -> (Option<JSON>, Vec<ApplyToError>);
+}
+
+impl ApplyToResultMethods for (Option<JSON>, Vec<ApplyToError>) {
+    // Intentionally taking ownership of self to avoid cloning, since we pretty
+    // much always use this method to replace the previous (value, errors) tuple
+    // before returning.
+    fn prepend_errors(self, errors: &[ApplyToError]) -> Self {
+        if errors.is_empty() {
+            self
+        } else {
+            let (value_opt, apply_errors) = self;
+            (value_opt, [errors, apply_errors.as_slice()].concat())
```

Comment:
I believe this is reallocating the error vec every single time even if we haven't accumulated any new errors. The good news is, if there are _no_ errors at all, I believe it's basically a no-op. I think, though, if we pass an owned `Vec<ApplyToError>` here and extend it we can save some cycles in the error case.

This should also make sure we aren't forgetting to pass around errors somewhere and doing extra cloning/duplicating, since when we pass ownership, the value is gone and we have to get a new one (the return value of this function)

Reply:
Improved this in https://github.com/apollographql/router/pull/6003/commits/37ab98d8159b9732223f88691d1703a1403bf762 and https://github.com/apollographql/router/pull/6003/commits/b04f54241bf4760b92a3c197d15b580f16ce149f, thanks!